Release/src/http/client/http_client_asio.cpp

```diff
@@ -354,9 +370,49 @@ class asio_connection_pool
 
     const int m_pool_timeout_secs;
     boost::asio::deadline_timer m_pool_timer;
-    int m_use_count;
 };
 
+asio_connection_pool_map &asio_connection_pool_map::instance()
+{
+    static asio_connection_pool_map s_instance;
+    return s_instance;
+}
+
+std::shared_ptr<asio_connection_pool> asio_connection_pool_map::obtain_connection_pool(const std::string &pool_key, bool start_with_ssl, std::chrono::seconds idle_timeout)
+{
+    std::lock_guard<std::mutex> lg(m_connection_pool_map_mutex);
+    auto &pool = m_connection_pool_map[pool_key];
+    if (!pool)
+    {
+        pool = std::make_shared<asio_connection_pool>(crossplat::threadpool::shared_instance().service(), start_with_ssl, idle_timeout, nullptr);
+    }
+
+    pool->cancel_pool_timer();
+    return pool;
+}
+
+void asio_connection_pool_map::release_connection_pool(const std::string &pool_key)
+{
+    std::lock_guard<std::mutex> lg(m_connection_pool_map_mutex);
+    auto &pool = m_connection_pool_map[pool_key];
+    if (pool)
+    {
+        pool->start_pool_timer(boost::bind(&asio_connection_pool_map::free_connection_pool, this, boost::asio::placeholders::error, pool_key));
+    }
+}
+
+void asio_connection_pool_map::free_connection_pool(const boost::system::error_code &ec, const std::string &pool_key)
+{
+    if (!ec)
+    {
+        std::lock_guard<std::mutex> lg(m_connection_pool_map_mutex);
+        auto &pool = m_connection_pool_map[pool_key];
```

Comment:
This will allocate a new pool `shared_ptr` if there is not a map entry. It would probably be better to use `.find(pool_key)` to avoid allocating a new map entry just to erase it.